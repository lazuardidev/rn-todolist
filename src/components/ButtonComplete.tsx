import React from 'react';
import {View, TouchableOpacity, StyleSheet} from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';

export const ButtonComplete = ({onPress}: {onPress: () => void}) => {
  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID="button-complete"
        style={styles.button}
        onPress={onPress}>
        <Ionicons name="checkmark-done" size={24} color="white" />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 120,
    right: 20,
  },
  button: {
    backgroundColor: 'purple',
    width: 60,
    height: 60,
    borderRadius: 30,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 3,
    shadowColor: 'black',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.3,
    shadowRadius: 2,
  },
});
