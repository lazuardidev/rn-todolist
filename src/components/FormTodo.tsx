import React from 'react';
import {View, StyleSheet, Text, TouchableOpacity} from 'react-native';
import {Button, TextInput, Checkbox} from 'react-native-paper';
import {TTodo} from '../utils/type';

type TFormTodo = {
  formData: TTodo;
  setFormData: React.Dispatch<React.SetStateAction<TTodo>>;
  onCreate?: () => void;
  onEdit?: (id: string) => void;
  onDelete?: (id: string) => void;
};

export const FormTodo = ({
  formData,
  setFormData,
  onCreate,
  onEdit,
  onDelete,
}: TFormTodo) => {
  return (
    <View style={styles.formContainer}>
      <TextInput
        testID="title"
        label="Title"
        mode="outlined"
        style={styles.input}
        placeholder="Title"
        value={formData.title}
        onChangeText={text => setFormData({...formData, title: text})}
      />
      <TextInput
        testID="description"
        label="Description"
        mode="outlined"
        style={styles.input}
        placeholder="Description"
        value={formData.description}
        multiline
        numberOfLines={16}
        onChangeText={text => setFormData({...formData, description: text})}
      />
      {onEdit && onDelete && (
        <TouchableOpacity
          style={styles.checkboxContainer}
          onPress={() =>
            setFormData({...formData, isComplete: !formData.isComplete})
          }>
          <Checkbox
            key={'mark-complete-checkbox'}
            status={formData.isComplete ? 'checked' : 'unchecked'}
            // onPress={() =>
            //   setFormData({...formData, isComplete: !formData.isComplete})
            // }
          />
          <Text>Mark as complete</Text>
        </TouchableOpacity>
      )}
      {onCreate && (
        <Button
          id="create-button"
          mode="contained"
          onPress={onCreate}
          style={styles.btn}>
          Create
        </Button>
      )}
      {onEdit && (
        <Button
          id="edit-button"
          mode="contained"
          onPress={() => onEdit(formData.id)}
          style={styles.btn}>
          Edit
        </Button>
      )}
      {onDelete && (
        <Button
          id="delete-button"
          mode="contained"
          buttonColor="red"
          onPress={() => onDelete(formData.id)}
          style={styles.btn}>
          Delete
        </Button>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  formContainer: {
    marginBottom: 20,
    width: '100%',
  },
  input: {
    marginBottom: 16,
  },
  btn: {
    width: '100%',
    marginVertical: 12,
  },
  checkboxContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 10,
  },
  checkbox: {
    alignSelf: 'center',
  },
});
