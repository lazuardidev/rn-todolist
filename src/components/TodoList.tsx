import React from 'react';
import {View, Text, StyleSheet, TouchableOpacity, FlatList} from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';

export const TodoList = ({
  data,
  onDeleteTodo,
  onToggleTodo,
  onNavigateToDetail,
}: any) => {
  const renderItem = ({item}: any) => {
    return (
      <TouchableOpacity
        onPress={() => onNavigateToDetail(item)}
        style={styles.itemContainer}>
        <View style={styles.rowContainer}>
          <Text
            testID={`title-${item.id}`}
            numberOfLines={1}
            ellipsizeMode="tail"
            style={styles.itemText}>
            {item.title}
          </Text>
          <TouchableOpacity>
            <View style={styles.buttonContainer}>
              <Ionicons
                testID={`btn-delete-${item.id}`}
                name="trash-outline"
                size={25}
                color="#333"
                onPress={() => onDeleteTodo(item.id)}
              />
              <Ionicons
                testID={`btn-complete-${item.id}`}
                name="checkmark-circle-outline"
                size={25}
                color="#333"
                onPress={() => onToggleTodo(item.id)}
              />
            </View>
          </TouchableOpacity>
        </View>
        <Text
          testID={`desc-${item.id}`}
          numberOfLines={2}
          ellipsizeMode="tail"
          style={styles.descText}>
          {item.description}
        </Text>
      </TouchableOpacity>
    );
  };

  return (
    <View>
      <FlatList
        testID="todo-list"
        data={data}
        renderItem={({item}) => renderItem({item})}
        keyExtractor={item => item.id}
      />
    </View>
  );
};

export default TodoList;

const styles = StyleSheet.create({
  itemContainer: {
    backgroundColor: 'white',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 8,
  },
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  buttonContainer: {
    display: 'flex',
    flexDirection: 'row',
    gap: 20,
  },
  itemText: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#333',
    maxWidth: 240,
  },
  descText: {
    fontSize: 14,
    color: '#333',
    marginTop: 16,
  },
  lineThrough: {
    textDecorationLine: 'line-through',
    fontSize: 16,
    color: '#333',
  },
});
