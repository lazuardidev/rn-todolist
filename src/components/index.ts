export * from './ButtonCreate';
export * from './ButtonComplete';
export * from './TodoList';
