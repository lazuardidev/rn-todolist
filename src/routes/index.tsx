import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import {SCREENS} from '../utils/constants';
import Home from '../pages/home/Home';
import Detail from '../pages/detail/Detail';
import Complete from '../pages/complete/Complete';
import Create from '../pages/create/Create';

const Stack = createStackNavigator();

const Routes = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator screenOptions={{headerShown: true}}>
        <Stack.Screen
          name={SCREENS.HOME}
          component={Home}
          options={{
            headerShown: false,
          }}
        />
        <Stack.Screen name={SCREENS.COMPLETE} component={Complete} />
        <Stack.Screen name={SCREENS.CREATE} component={Create} />
        <Stack.Screen name={SCREENS.DETAIL} component={Detail} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default Routes;
