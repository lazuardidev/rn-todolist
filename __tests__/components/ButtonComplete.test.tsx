import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {ButtonComplete} from '../../src/components';

describe('ButtonComplete', () => {
  it('should call onPress when button is pressed', () => {
    const onPressMock = jest.fn();
    const {getByTestId} = render(<ButtonComplete onPress={onPressMock} />);
    fireEvent.press(getByTestId('button-complete'));
    expect(onPressMock).toHaveBeenCalled();
  });

  it('should render correctly', () => {
    const onPressMock = jest.fn();
    const {getByTestId} = render(<ButtonComplete onPress={onPressMock} />);
    expect(getByTestId('button-complete')).toBeTruthy();
  });
});
