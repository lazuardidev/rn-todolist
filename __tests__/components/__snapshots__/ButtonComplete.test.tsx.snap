// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComplete should render correctly 1`] = `ShallowWrapper {}`;
